### ISSUE TITLE
If a child widget changes the virtual size of a parent, the scrollbars don't update immediately.

### ISSUE THREAD

**Issue by willmcgugan** [Textualize/textual#1166]

**Comment by willmcgugan** [comment]
@darrenburns Do you recall the exact arrangement that causes this issue?

Tried to reproduce with a Static that increases its size. Seems to work, in that the scrollbar updates as I'd expect.

**Comment by darrenburns** [comment]
The issue was more that if you scroll a container say half way down, then a render occurs with new content, the scrollbar stays at the half way point but the new content isn't scrolled.

Hopefully that makes sense, hard to explain. I'll try to make an example of it if not.

**Comment by aaronst** [comment]
Wondering if #1215 is at all related.

**Comment by willmcgugan** [comment]
@Textualize/dev Who wants to investigate this one? It might require a bit of digging in to the internals.

**Comment by davep** [comment]
There seems to have been some issue with coming up with a nice way of reproducing this on demand, but as it turns out I seem to have managed with [PISpy](https://github.com/davep/pispy). If you have a watch over [this video of it in action](https://www.youtube.com/watch?v=yMGD6bXqIEo) you can nicely see that, when I click on a requirement link near the bottom of a display, the next display starts at the top of all the data (as wanted) but the scrollbar, and the container's apparent knowledge of scroll position seems to be off. The moment I touch the mouse to interact it jumps to where it thinks it should be.

Point being, with the right combination of packages to look at (`ipython` seems to be a good one to start with), I think there's a good method to reproduce here (and from this I think it now should be easy enough to work out a standalone example).

Recipe from this is to have a `Vertical` be populated with lots of widgets such that scrolling happens. then `self.remove( "ThisVerticalDerivedContainer > *" ).remove`, followed by populating with a different number of child widgets, ideally that still cause scroll -- all done after scrolling to the bottom.

**Comment by davep** [comment]
Following on from the above, I had hoped that this would isolate the problem, making it a nicely self-contained example:

```python
from textual.app import App, ComposeResult
from textual.containers import Vertical
from textual.widget import Widget
from textual.widgets import Header, Footer, Label

class ScrollApp( App[ None ] ):

    BINDINGS = [
        ( "d", "decimate", "Decimate" )
    ]

    def widgets( self, count: int ) -> list[ Widget ]:
        return [
            Label( f"This is label {n}", id=f"line{n}" ) for n in range( count )
        ]

    def compose(self) -> ComposeResult:
        yield Header()
        yield Vertical( *self.widgets( 500 ) )
        yield Footer()

    def action_decimate( self ) -> None:
        widgets = self.query( "Vertical > *" )
        new_count = int( len( widgets ) * 0.9 )
        widgets.remove()
        self.query_one( Vertical ).mount( *self.widgets( new_count ) )

if __name__ == "__main__":
    ScrollApp().run()
```

Sadly though, no matter where I scroll to before hitting `d`, I can't reproduce the issue. **However**, if I sprinkle in a little `async` and `await` into `action_decimate` (and in PISpy it's some `async`/`await` code involved)...

```python
from textual.app import App, ComposeResult
from textual.containers import Vertical
from textual.widget import Widget
from textual.widgets import Header, Footer, Label

class ScrollApp( App[ None ] ):

    BINDINGS = [
        ( "d", "decimate", "Decimate" )
    ]

    def widgets( self, count: int ) -> list[ Widget ]:
        return [
            Label( f"This is label {n}", id=f"line{n}" ) for n in range( count )
        ]

    def compose(self) -> ComposeResult:
        yield Header()
        yield Vertical( *self.widgets( 500 ) )
        yield Footer()

    async def action_decimate( self ) -> None:
        widgets = self.query( "Vertical > *" )
        new_count = int( len( widgets ) * 0.9 )
        await widgets.remove()
        self.query_one( Vertical ).mount( *self.widgets( new_count ) )

if __name__ == "__main__":
    ScrollApp().run()
```

Now I can reproduce the issue. It doesn't always happen on the first hit of `d`, but it's rare that it doesn't start to show after the second or third press.

Note that for best effect you should scroll the `Vertical` to the very bottom before hitting `d`.

**Comment by willmcgugan** [comment]
@davep Not 100% this is the same issue. I find your description a little hard to follow. Any chance you could narrow it down a little, and post as an issue?

**Comment by davep** [comment]
@willmcgugan The above (latter) code was as narrow as I could get it; if it's a different issue I'm happy enough to raise one for it.

**Comment by willmcgugan** [comment]
Will go over layout system tomorrow, and tackle it after.

**Comment by darrenburns** [comment]
> Not 100% this is the same issue. I find your description a little hard to follow. Any chance you could narrow it down a little, and post as an issue?

The issue Dave reported is the same as what I was experiencing here, so unless there's another problem I'm not aware of, I don't think this issue is required any more.
